chat-android/src/main/java/com/ably/chat/Utils.kt

```diff
@@ -217,3 +229,69 @@ internal class LatestJobExecutor {
         }
     }
 }
+
+/**
+ * A custom implementation of a `MutableSharedFlow` that supports optional awaitable emissions.
+ * This class allows emitting values with an optional `CompletableDeferred` to track the completion
+ * of processing for each emitted value. It is useful in scenarios where you need to ensure that
+ * all consumers have processed the emitted values before proceeding.
+ * Note - Only one collector is allowed to process events.
+ */
+internal class AwaitableSharedFlow<T>(private val logger: Logger) {
```

Comment:
I have some concerns about the public API design. 

The current approach where we continuously emit values but occasionally want to wait for pending events seems a bit counterintuitive. Since events are processed sequentially by a single collector, there's no actual need to maintain a collection of all completion deferreds. Only the most recent one would be necessary to know when all pending events have been processed.

Perhaps we could simplify this by redesigning the API to make the sequential nature of the processing more explicit and provide a cleaner way to await completion when needed (also we can consider using Channel to implement this, since SharedFlow usually doesn't fit well with single producer single consumer pattern)

Reply:
Hmm.. using channels seems like a good idea 🤔. 
But does it allows to publish from non suspended method, need to check

Reply:
Actually, it was intentional to wait for only those events that were emitted as a part of pubsub attach, detach operation. If we try to wait on last processed event, there's a chance we wait on events that are not emitted as a part of attach or detach. 
But this feels more like an edge case.

Reply:
https://github.com/ably/ably-chat-kotlin/pull/124/commits/b26bc37d23daa45b55080bf681418f4baf3a7fc1